Source/Widgets/Geocoder/GeocoderViewModel.js

```diff
@@ -225,81 +399,128 @@ define([
         });
     }
 
-    function geocode(viewModel) {
-        var query = viewModel.searchText;
+    function chainPromise(promise, geocoderService, query) {
+        return promise
+            .then(function(result) {
+                if (defined(result) && result.state === 'fulfilled' && result.value.length > 0){
+                    return result;
+                }
+                var nextPromise = geocoderService.geocode(query)
+                    .then(function (result) {
+                        return {state: 'fulfilled', value: result};
+                    });
+                if (defined(nextPromise.otherwise)) {
+                    nextPromise.otherwise(function (err) {
+                        return {state: 'rejected', reason: err};
+                    });
+                } else if (defined(nextPromise.catch)) {
+                    nextPromise.catch(function (err) {
+                        return {state: 'rejected', reason: err};
+                    });
+                }
+                return nextPromise;
+            });
+    }
 
-        if (/^\s*$/.test(query)) {
-            //whitespace string
+    function geocode(viewModel, geocoderServices) {
+        var query = viewModel._searchText;
+
+        if (hasOnlyWhitespace(query)) {
             return;
         }
 
-        // If the user entered (longitude, latitude, [height]) in degrees/meters,
-        // fly without calling the geocoder.
-        var splitQuery = query.match(/[^\s,\n]+/g);
-        if ((splitQuery.length === 2) || (splitQuery.length === 3)) {
-            var longitude = +splitQuery[0];
-            var latitude = +splitQuery[1];
-            var height = (splitQuery.length === 3) ? +splitQuery[2] : 300.0;
-
-            if (!isNaN(longitude) && !isNaN(latitude) && !isNaN(height)) {
-                updateCamera(viewModel, Cartesian3.fromDegrees(longitude, latitude, height));
-                return;
-            }
-        }
         viewModel._isSearchInProgress = true;
 
-        var promise = loadJsonp(viewModel._url + 'REST/v1/Locations', {
-            parameters : {
-                query : query,
-                key : viewModel._key
+        var promise = when.resolve();
+        for (var i = 0; i < geocoderServices.length; i++) {
+            promise = chainPromise(promise, geocoderServices[i], query);
+        }
 
-            },
-            callbackParameterName : 'jsonp'
-        });
+        viewModel._geocodePromise = promise;
+        promise
+            .then(function (result) {
+                if (promise.cancel || promise !== viewModel._geocodePromise) {
```

Comment:
True.  I was thinking it would happen if you hit the geocode button a few times, but in that case `promise.cancel` would be true.  I just kept this here because that's how it was working before, but I agree that it doesn't seem necessary.